ykrt/src/compile/jitc_yk/codegen/x64/lsregalloc.rs

```diff
@@ -435,60 +437,72 @@ impl LSRegAlloc<'_> {
-        // Note: during this stage no changes are made to the allocator's state in `self`.
-
-        let (mut asgn_regs, cnstr_regs) = self.find_regs_for_constraints(iidx, &mut constraints);
+        // Stage 1: Find the register we will use for each `GPConstraint`. Note: we may use the
+        // same register for multiple constraints.
 
-        // Stage 2: At this point, we've found a register for every constraint. We now go about
-        // updating the allocator's state to get from where we were to where we want to be.
+        let (asgn_regs, cnstr_regs) = self.find_regs_for_constraints(iidx, &cnstrs);
+        assert_eq!(
+            cnstr_regs
+                .iter()
+                .filter(|reg| asgn_regs.is_set(**reg))
+                .count(),
+            cnstr_regs.len()
+        );
 
-        // Spill currently assigned registers that don't contain values we want for the current
-        // instruction.
-        self.spill_or_move_constraints(asm, iidx, &constraints, &mut asgn_regs, &cnstr_regs);
+        // Stage 2: At this point, we've found a register for every constraint. Decide how we will
+        // make use of values in existing registers. Note: this stage does not update any state in
+        // `self`.
+
+        // Stage 2.1: For every operand that is in a register -- even if not the right register! --
+        // generate copies for it.
+        let copies = self.input_regs_to_copies(iidx, &cnstrs, &cnstr_regs);
+        let actions = reg_copies_to_actions(copies);
+
+        // Stage 2.2: the copies that we must do above may end up clobbering existing registers.
+        // If those registers contain values we still need, move them to another register (if there
+        // are spare/empty registers) or otherwise ensure those values are spilled.
+        let actions = self.move_or_spill_clobbered_regs(iidx, &actions, &cnstrs, &cnstr_regs);
+
+        // Stage 3: Generate code and update `self`.
+
+        // Stage 3.1: Go through the sequence of actions, copying / spilling values.
+        for (reg, action) in &actions {
+            match action {
+                RegAction::Keep => (),
+                RegAction::CopyFrom(from_reg) => {
+                    assert_ne!(from_reg, reg);
+                    dynasm!(asm; mov Rq(reg.code()), Rq(from_reg.code()));
+                    self.gp_regset.set(*reg);
+                    let st = self.gp_reg_states[usize::from(from_reg.code())].clone();
+                    self.gp_reg_states[usize::from(reg.code())] = st;
+                }
+                RegAction::Spill => {
+                    self.spill_gp_if_not_already(asm, iidx, *reg);
+                }
+            }
+        }
 
-        // Put values that aren't in registers into them -- spilling values if we need them later.
-        for (cnstr, reg) in constraints.iter().zip(cnstr_regs.into_iter()) {
+        // Stage 3.2: For any input constraints that don't yet contain a value, unspill the value.
```

Comment:
I don't fully understand this step. Does this unspill values that we have just spilled in 3.1?

Reply:
No: the previous steps have got all values that were into registers into the right registers. Any registers that don't yet have the right value must have been spilled by a previous instruction.